pkg/volume/gce_pd/attacher.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce_pd
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type gcePersistentDiskAttacher struct{}
+
+var _ volume.Attacher = &gcePersistentDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &gcePersistentDiskPlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the same PD atomic
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *gcePersistentDiskPlugin) NewAttacher() (volume.Attacher, error) {
+	return &gcePersistentDiskAttacher{}, nil
+}
+
+func (attacher *gcePersistentDiskAttacher) Attach(cloudProvider cloudprovider.Interface, spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	pdName := volumeSource.PDName
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(pdName)
+	defer attachDetachMutex.UnlockKey(pdName)
+
+	gceCloud, err := getCloudProvider(cloudProvider)
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		var attached bool
+		attached, err = gceCloud.DiskIsAttached(pdName, hostName)
```

Comment:
> We can't do this (succeed if volume is already) yet. If the disk is attached and you succeed here, since the volume attach/mount and detach/unmount loops are still async a bunch of race conditions that are hidden will manifest them selves.

Oh yes; good catch.. I forgot about that scenario :/



> By the way the test/e2e/pd.go are not run as the default set on PRs, so please run them manually to confirm they are passing (they are part of the slow suite, so after PR is merged they could break and block the submit queue).

Yup.. I have been running that test regularly as part of this work.



> A good way to catch some of the races is to run theses tests back to back for 30min to 1hr.

That I have not done.. will do